torch/_ops.py

```diff
@@ -409,6 +430,93 @@ def key_extractor(tensors, key_mask):
 _mode_stack_per_key: Dict[torch._C.DispatchKey, List] = {}
 
 
+class ModeSlot(enum.IntEnum):
+    PROXY_MODE_SLOT = 0
+    FUNCTIONAL_MODE_SLOT = 1
+
+
+_mode_stack_per_key[torch._C.DispatchKey.PreDispatch] = [None, None]
+
+
+def _get_pre_dispatch_slot(mode_key):
+    assert mode_key in (
+        torch._C._TorchDispatchModeKey.PROXY,
+        torch._C._TorchDispatchModeKey.FUNCTIONAL,
+    )
+    if mode_key == torch._C._TorchDispatchModeKey.PROXY:
+        return ModeSlot.PROXY_MODE_SLOT
+    return ModeSlot.FUNCTIONAL_MODE_SLOT
+
+
+def unset_mode_pre_dispatch(mode_key):
+    current_mode_stack_pre_dispatch = mode_stack_per_key()[
+        torch._C.DispatchKey.PreDispatch
+    ]
+    assert mode_key in (
+        torch._C._TorchDispatchModeKey.PROXY,
+        torch._C._TorchDispatchModeKey.FUNCTIONAL,
+    )
+    slot = _get_pre_dispatch_slot(mode_key)
+    current_mode = current_mode_stack_pre_dispatch[slot]
+    mode_stack_per_key()[torch._C.DispatchKey.PreDispatch][slot] = None
+    return current_mode
+
+
+def _set_mode_pre_dispatch(mode):
+    assert hasattr(mode, "_mode_key") and mode._mode_key in (
+        torch._C._TorchDispatchModeKey.PROXY,
+        torch._C._TorchDispatchModeKey.FUNCTIONAL,
+    )
+    mode_key = mode._mode_key
+    slot = _get_pre_dispatch_slot(mode_key)
+    current_mode = mode_stack_per_key()[torch._C.DispatchKey.PreDispatch][slot]
+    assert current_mode is None
+    mode_stack_per_key()[torch._C.DispatchKey.PreDispatch][slot] = mode
+
+
+def _pop_mode_from_pre_dispatch():
+    for i in range(1, -1, -1):
+        current_mode = mode_stack_per_key()[torch._C.DispatchKey.PreDispatch][i]
+        if current_mode is not None:
+            mode_stack_per_key()[torch._C.DispatchKey.PreDispatch][i] = None
+            return current_mode
+    raise AssertionError("Trying to pop empty mode stack")
+
+
+def _len_torch_dispatch_stack_pre_dispatch():
+    count = 0
+    for mode in mode_stack_per_key()[torch._C.DispatchKey.PreDispatch]:
+        if mode is not None:
+            count += 1
```

Comment:
hmm I don't love that we have a "list" that can have None entries in it, so you need to know to indirect through this function  (so it's wrong to accidentally grab the actual list and look at it).

To make it more explicit, what if we just don't use a list at all to hold the two modes, and have two private slots (_proxy_slot, _functional_slot), that all of these helpers check?